File: PHP/src/OneLang/One/AstTransformer.php
<?php

namespace OneLang\One\AstTransformer;

use OneLang\One\Ast\AstTypes\IHasTypeArguments;
use OneLang\One\Ast\AstTypes\ClassType;
use OneLang\One\Ast\AstTypes\InterfaceType;
use OneLang\One\Ast\AstTypes\UnresolvedType;
use OneLang\One\Ast\AstTypes\LambdaType;
use OneLang\One\Ast\Expressions\Identifier;
use OneLang\One\Ast\Expressions\BinaryExpression;
use OneLang\One\Ast\Expressions\ConditionalExpression;
use OneLang\One\Ast\Expressions\NewExpression;
use OneLang\One\Ast\Expressions\TemplateString;
use OneLang\One\Ast\Expressions\ParenthesizedExpression;
use OneLang\One\Ast\Expressions\UnaryExpression;
use OneLang\One\Ast\Expressions\PropertyAccessExpression;
use OneLang\One\Ast\Expressions\ElementAccessExpression;
use OneLang\One\Ast\Expressions\ArrayLiteral;
use OneLang\One\Ast\Expressions\MapLiteral;
use OneLang\One\Ast\Expressions\Expression;
use OneLang\One\Ast\Expressions\CastExpression;
use OneLang\One\Ast\Expressions\UnresolvedCallExpression;
use OneLang\One\Ast\Expressions\InstanceOfExpression;
use OneLang\One\Ast\Expressions\AwaitExpression;
use OneLang\One\Ast\Expressions\StringLiteral;
use OneLang\One\Ast\Expressions\NumericLiteral;
use OneLang\One\Ast\Expressions\NullLiteral;
use OneLang\One\Ast\Expressions\RegexLiteral;
use OneLang\One\Ast\Expressions\BooleanLiteral;
use OneLang\One\Ast\Expressions\StaticMethodCallExpression;
use OneLang\One\Ast\Expressions\InstanceMethodCallExpression;
use OneLang\One\Ast\Expressions\UnresolvedNewExpression;
use OneLang\One\Ast\Expressions\NullCoalesceExpression;
use OneLang\One\Ast\Expressions\UnresolvedMethodCallExpression;
use OneLang\One\Ast\Expressions\GlobalFunctionCallExpression;
use OneLang\One\Ast\Expressions\LambdaCallExpression;
use OneLang\One\Ast\Statements\ReturnStatement;
use OneLang\One\Ast\Statements\ExpressionStatement;
use OneLang\One\Ast\Statements\IfStatement;
use OneLang\One\Ast\Statements\ThrowStatement;
use OneLang\One\Ast\Statements\VariableDeclaration;
use OneLang\One\Ast\Statements\WhileStatement;
use OneLang\One\Ast\Statements\ForStatement;
use OneLang\One\Ast\Statements\ForeachStatement;
use OneLang\One\Ast\Statements\Statement;
use OneLang\One\Ast\Statements\UnsetStatement;
use OneLang\One\Ast\Statements\BreakStatement;
use OneLang\One\Ast\Statements\ContinueStatement;
use OneLang\One\Ast\Statements\DoStatement;
use OneLang\One\Ast\Statements\TryStatement;
use OneLang\One\Ast\Statements\Block;
use OneLang\One\Ast\Types\Method;
use OneLang\One\Ast\Types\Constructor;
use OneLang\One\Ast\Types\Field;
use OneLang\One\Ast\Types\Property;
use OneLang\One\Ast\Types\Interface_;
use OneLang\One\Ast\Types\Class_;
use OneLang\One\Ast\Types\Enum;
use OneLang\One\Ast\Types\EnumMember;
use OneLang\One\Ast\Types\SourceFile;
use OneLang\One\Ast\Types\IVariable;
use OneLang\One\Ast\Types\IVariableWithInitializer;
use OneLang\One\Ast\Types\MethodParameter;
use OneLang\One\Ast\Types\Lambda;
use OneLang\One\Ast\Types\IMethodBase;
use OneLang\One\Ast\Types\Package;
use OneLang\One\Ast\Types\GlobalFunction;
use OneLang\One\Ast\Types\IInterface;
use OneLang\One\Ast\Types\IAstNode;
use OneLang\One\Ast\Types\IHasAttributesAndTrivia;
use OneLang\One\Ast\Types\Import;
use OneLang\One\Ast\References\ClassReference;
use OneLang\One\Ast\References\EnumReference;
use OneLang\One\Ast\References\ThisReference;
use OneLang\One\Ast\References\MethodParameterReference;
use OneLang\One\Ast\References\VariableDeclarationReference;
use OneLang\One\Ast\References\ForVariableReference;
use OneLang\One\Ast\References\ForeachVariableReference;
use OneLang\One\Ast\References\SuperReference;
use OneLang\One\Ast\References\InstanceFieldReference;
use OneLang\One\Ast\References\InstancePropertyReference;
use OneLang\One\Ast\References\StaticPropertyReference;
use OneLang\One\Ast\References\StaticFieldReference;
use OneLang\One\Ast\References\CatchVariableReference;
use OneLang\One\Ast\References\GlobalFunctionReference;
use OneLang\One\Ast\References\EnumMemberReference;
use OneLang\One\Ast\References\StaticThisReference;
use OneLang\One\Ast\References\VariableReference;
use OneLang\One\ErrorManager\ErrorManager;
use OneLang\One\ITransformer\ITransformer;
use OneLang\One\Ast\Interfaces\IType;

class AstTransformer implements ITransformer {
    public $errorMan;
    public $currentFile;
    public $currentInterface;
    public $currentMethod;
    public $currentClosure;
    public $currentStatement;
    public $name;
    
    function __construct($name) {
        $this->name = $name;
        $this->errorMan = new ErrorManager();
        $this->currentFile = null;
        $this->currentInterface = null;
        $this->currentMethod = null;
        $this->currentClosure = null;
        $this->currentStatement = null;
    }
    
    protected function visitAttributesAndTrivia($node) {
        
    }
    
    protected function visitType($type) {
        if ($type instanceof ClassType || $type instanceof InterfaceType || $type instanceof UnresolvedType) {
            $type2 = $type;
            $type2->typeArguments = array_map(function ($x) { return $this->visitType($x); }, $type2->typeArguments);
        }
        else if ($type instanceof LambdaType) {
            foreach ($type->parameters as $mp)
                $this->visitMethodParameter($mp);
            $type->returnType = $this->visitType($type->returnType);
        }
        return $type;
    }
    
    protected function visitIdentifier($id) {
        return $id;
    }
    
    protected function visitVariable($variable) {
        if ($variable->type !== null)
            $variable->type = $this->visitType($variable->type);
        return $variable;
    }
    
    protected function visitVariableWithInitializer($variable) {
        $this->visitVariable($variable);
        if ($variable->initializer !== null)
            $variable->initializer = $this->visitExpression($variable->initializer);
        return $variable;
    }
    
    protected function visitVariableDeclaration($stmt) {
        $this->visitVariableWithInitializer($stmt);
        return $stmt;
    }
    
    protected function visitUnknownStatement($stmt) {
        $this->errorMan->throw("Unknown statement type");
        return $stmt;
    }
    
    protected function visitStatement($stmt) {
        $this->currentStatement = $stmt;
        $this->visitAttributesAndTrivia($stmt);
        if ($stmt instanceof ReturnStatement) {
            if ($stmt->expression !== null)
                $stmt->expression = $this->visitExpression($stmt->expression);
        }
        else if ($stmt instanceof ExpressionStatement)
            $stmt->expression = $this->visitExpression($stmt->expression);
        else if ($stmt instanceof IfStatement) {
            $stmt->condition = $this->visitExpression($stmt->condition);
            $stmt->then = $this->visitBlock($stmt->then);
            if ($stmt->else_ !== null)
                $stmt->else_ = $this->visitBlock($stmt->else_);
        }
        else if ($stmt instanceof ThrowStatement)
            $stmt->expression = $this->visitExpression($stmt->expression);
        else if ($stmt instanceof VariableDeclaration)
            return $this->visitVariableDeclaration($stmt);
        else if ($stmt instanceof WhileStatement) {
            $stmt->condition = $this->visitExpression($stmt->condition);
            $stmt->body = $this->visitBlock($stmt->body);
        }
        else if ($stmt instanceof DoStatement) {
            $stmt->condition = $this->visitExpression($stmt->condition);
            $stmt->body = $this->visitBlock($stmt->body);
        }
        else if ($stmt instanceof ForStatement) {
            if ($stmt->itemVar !== null)
                $this->visitVariableWithInitializer($stmt->itemVar);
            $stmt->condition = $this->visitExpression($stmt->condition);
            $stmt->incrementor = $this->visitExpression($stmt->incrementor);
            $stmt->body = $this->visitBlock($stmt->body);
        }
        else if ($stmt instanceof ForeachStatement) {
            $this->visitVariable($stmt->itemVar);
            $stmt->items = $this->visitExpression($stmt->items);
            $stmt->body = $this->visitBlock($stmt->body);
        }
        else if ($stmt instanceof TryStatement) {
            $stmt->tryBody = $this->visitBlock($stmt->tryBody);
            if ($stmt->catchBody !== null) {
                $this->visitVariable($stmt->catchVar);
                $stmt->catchBody = $this->visitBlock($stmt->catchBody);
            }
            if ($stmt->finallyBody !== null)
                $stmt->finallyBody = $this->visitBlock($stmt->finallyBody);
        }
        else if ($stmt instanceof BreakStatement) { }
        else if ($stmt instanceof UnsetStatement)
            $stmt->expression = $this->visitExpression($stmt->expression);
        else if ($stmt instanceof ContinueStatement) { }
        else
            return $this->visitUnknownStatement($stmt);
        return $stmt;
    }
    
    protected function visitBlock($block) {
        $block->statements = array_map(function ($x) { return $this->visitStatement($x); }, $block->statements);
        return $block;
    }
    
    protected function visitTemplateString($expr) {
        for ($i = 0; $i < count($expr->parts); $i++) {
            $part = $expr->parts[$i];
            if (!$part->isLiteral)
                $part->expression = $this->visitExpression($part->expression);
        }
        return $expr;
    }
    
    protected function visitUnknownExpression($expr) {
        $this->errorMan->throw("Unknown expression type");
        return $expr;
    }
    
    protected function visitLambda($lambda) {
        $prevClosure = $this->currentClosure;
        $this->currentClosure = $lambda;
        $this->visitMethodBase($lambda);
        $this->currentClosure = $prevClosure;
        return $lambda;
    }
    
    protected function visitVariableReference($varRef) {
        return $varRef;
    }
    
    protected function visitExpression($expr) {
        if ($expr instanceof BinaryExpression) {
            $expr->left = $this->visitExpression($expr->left);
            $expr->right = $this->visitExpression($expr->right);
        }
        else if ($expr instanceof NullCoalesceExpression) {
            $expr->defaultExpr = $this->visitExpression($expr->defaultExpr);
            $expr->exprIfNull = $this->visitExpression($expr->exprIfNull);
        }
        else if ($expr instanceof UnresolvedCallExpression) {
            $expr->func = $this->visitExpression($expr->func);
            $expr->typeArgs = array_map(function ($x) { return $this->visitType($x); }, $expr->typeArgs);
            $expr->args = array_map(function ($x) { return $this->visitExpression($x); }, $expr->args);
        }
        else if ($expr instanceof UnresolvedMethodCallExpression) {
            $expr->object = $this->visitExpression($expr->object);
            $expr->typeArgs = array_map(function ($x) { return $this->visitType($x); }, $expr->typeArgs);
            $expr->args = array_map(function ($x) { return $this->visitExpression($x); }, $expr->args);
        }
        else if ($expr instanceof ConditionalExpression) {
            $expr->condition = $this->visitExpression($expr->condition);
            $expr->whenTrue = $this->visitExpression($expr->whenTrue);
            $expr->whenFalse = $this->visitExpression($expr->whenFalse);
        }
        else if ($expr instanceof Identifier)
            return $this->visitIdentifier($expr);
        else if ($expr instanceof UnresolvedNewExpression) {
            $this->visitType($expr->cls);
            $expr->args = array_map(function ($x) { return $this->visitExpression($x); }, $expr->args);
        }
        else if ($expr instanceof NewExpression) {
            $this->visitType($expr->cls);
            $expr->args = array_map(function ($x) { return $this->visitExpression($x); }, $expr->args);
        }
        else if ($expr instanceof TemplateString)
            return $this->visitTemplateString($expr);
        else if ($expr instanceof ParenthesizedExpression)
            $expr->expression = $this->visitExpression($expr->expression);
        else if ($expr instanceof UnaryExpression)
            $expr->operand = $this->visitExpression($expr->operand);
        else if ($expr instanceof PropertyAccessExpression)
            $expr->object = $this->visitExpression($expr->object);
        else if ($expr instanceof ElementAccessExpression) {
            $expr->object = $this->visitExpression($expr->object);
            $expr->elementExpr = $this->visitExpression($expr->elementExpr);
        }
        else if ($expr instanceof ArrayLiteral)
            $expr->items = array_map(function ($x) { return $this->visitExpression($x); }, $expr->items);
        else if ($expr instanceof MapLiteral)
            foreach ($expr->items as $item)
                $item->value = $this->visitExpression($item->value);
        else if ($expr instanceof StringLiteral) { }
        else if ($expr instanceof BooleanLiteral) { }
        else if ($expr instanceof NumericLiteral) { }
        else if ($expr instanceof NullLiteral) { }
        else if ($expr instanceof RegexLiteral) { }
        else if ($expr instanceof CastExpression) {
            $expr->newType = $this->visitType($expr->newType);
            $expr->expression = $this->visitExpression($expr->expression);
        }
        else if ($expr instanceof InstanceOfExpression) {
            $expr->expr = $this->visitExpression($expr->expr);
            $expr->checkType = $this->visitType($expr->checkType);
        }
        else if ($expr instanceof AwaitExpression)
            $expr->expr = $this->visitExpression($expr->expr);
        else if ($expr instanceof Lambda)
            return $this->visitLambda($expr);
        else if ($expr instanceof ClassReference) { }
        else if ($expr instanceof EnumReference) { }
        else if ($expr instanceof ThisReference) { }
        else if ($expr instanceof StaticThisReference) { }
        else if ($expr instanceof MethodParameterReference)
            return $this->visitVariableReference($expr);
        else if ($expr instanceof VariableDeclarationReference)
            return $this->visitVariableReference($expr);
        else if ($expr instanceof ForVariableReference)
            return $this->visitVariableReference($expr);
        else if ($expr instanceof ForeachVariableReference)
            return $this->visitVariableReference($expr);
        else if ($expr instanceof CatchVariableReference)
            return $this->visitVariableReference($expr);
        else if ($expr instanceof GlobalFunctionReference) { }
        else if ($expr instanceof SuperReference) { }
        else if ($expr instanceof InstanceFieldReference) {
            $expr->object = $this->visitExpression($expr->object);
            return $this->visitVariableReference($expr);
        }
        else if ($expr instanceof InstancePropertyReference) {
            $expr->object = $this->visitExpression($expr->object);
            return $this->visitVariableReference($expr);
        }
        else if ($expr instanceof StaticFieldReference)
            return $this->visitVariableReference($expr);
        else if ($expr instanceof StaticPropertyReference)
            return $this->visitVariableReference($expr);
        else if ($expr instanceof EnumMemberReference) { }
        else if ($expr instanceof StaticMethodCallExpression) {
            $expr->typeArgs = array_map(function ($x) { return $this->visitType($x); }, $expr->typeArgs);
            $expr->args = array_map(function ($x) { return $this->visitExpression($x); }, $expr->args);
        }
        else if ($expr instanceof GlobalFunctionCallExpression)
            $expr->args = array_map(function ($x) { return $this->visitExpression($x); }, $expr->args);
        else if ($expr instanceof InstanceMethodCallExpression) {
            $expr->object = $this->visitExpression($expr->object);
            $expr->typeArgs = array_map(function ($x) { return $this->visitType($x); }, $expr->typeArgs);
            $expr->args = array_map(function ($x) { return $this->visitExpression($x); }, $expr->args);
        }
        else if ($expr instanceof LambdaCallExpression)
            $expr->args = array_map(function ($x) { return $this->visitExpression($x); }, $expr->args);
        else
            return $this->visitUnknownExpression($expr);
        return $expr;
    }
    
    protected function visitMethodParameter($methodParameter) {
        $this->visitAttributesAndTrivia($methodParameter);
        $this->visitVariableWithInitializer($methodParameter);
    }
    
    protected function visitMethodBase($method) {
        foreach ($method->parameters as $item)
            $this->visitMethodParameter($item);
        
        if ($method->body !== null)
            $method->body = $this->visitBlock($method->body);
    }
    
    protected function visitMethod($method) {
        $this->currentMethod = $method;
        $this->currentClosure = $method;
        $this->visitAttributesAndTrivia($method);
        $this->visitMethodBase($method);
        $method->returns = $this->visitType($method->returns);
        $this->currentClosure = null;
        $this->currentMethod = null;
    }
    
    protected function visitGlobalFunction($func) {
        $this->visitMethodBase($func);
        $func->returns = $this->visitType($func->returns);
    }
    
    protected function visitConstructor($constructor) {
        $this->currentMethod = $constructor;
        $this->currentClosure = $constructor;
        $this->visitAttributesAndTrivia($constructor);
        $this->visitMethodBase($constructor);
        $this->currentClosure = null;
        $this->currentMethod = null;
    }
    
    protected function visitField($field) {
        $this->visitAttributesAndTrivia($field);
        $this->visitVariableWithInitializer($field);
    }
    
    protected function visitProperty($prop) {
        $this->visitAttributesAndTrivia($prop);
        $this->visitVariable($prop);
        if ($prop->getter !== null)
            $prop->getter = $this->visitBlock($prop->getter);
        if ($prop->setter !== null)
            $prop->setter = $this->visitBlock($prop->setter);
    }
    
    protected function visitInterface($intf) {
        $this->currentInterface = $intf;
        $this->visitAttributesAndTrivia($intf);
        $intf->baseInterfaces = array_map(function ($x) { return $this->visitType($x); }, $intf->baseInterfaces);
        foreach ($intf->fields as $field)
            $this->visitField($field);
        foreach ($intf->methods as $method)
            $this->visitMethod($method);
        $this->currentInterface = null;
    }
    
    protected function visitClass($cls) {
        $this->currentInterface = $cls;
        $this->visitAttributesAndTrivia($cls);
        if ($cls->constructor_ !== null)
            $this->visitConstructor($cls->constructor_);
        
        $cls->baseClass = $this->visitType($cls->baseClass);
        $cls->baseInterfaces = array_map(function ($x) { return $this->visitType($x); }, $cls->baseInterfaces);
        foreach ($cls->fields as $field)
            $this->visitField($field);
        foreach ($cls->properties as $prop)
            $this->visitProperty($prop);
        foreach ($cls->methods as $method)
            $this->visitMethod($method);
        $this->currentInterface = null;
    }
    
    protected function visitEnum($enum_) {
        $this->visitAttributesAndTrivia($enum_);
        foreach ($enum_->values as $value)
            $this->visitEnumMember($value);
    }
    
    protected function visitEnumMember($enumMember) {
        
    }
    
    protected function visitImport($imp) {
        $this->visitAttributesAndTrivia($imp);
    }
    
    function visitFile($sourceFile) {
        $this->errorMan->resetContext($this);
        $this->currentFile = $sourceFile;
        foreach ($sourceFile->imports as $imp)
            $this->visitImport($imp);
        foreach ($sourceFile->enums as $enum_)
            $this->visitEnum($enum_);
        foreach ($sourceFile->interfaces as $intf)
            $this->visitInterface($intf);
        foreach ($sourceFile->classes as $cls)
            $this->visitClass($cls);
        foreach ($sourceFile->funcs as $func)
            $this->visitGlobalFunction($func);
        $sourceFile->mainBlock = $this->visitBlock($sourceFile->mainBlock);
        $this->currentFile = null;
    }
    
    function visitFiles($files) {
        foreach ($files as $file)
            $this->visitFile($file);
    }
    
    function visitPackage($pkg) {
        $this->visitFiles(array_values($pkg->files));
    }
}

File: Python/OneLang/One/AstTransformer.py
from onelang_core import *
import OneLang.One.Ast.AstTypes as astTypes
import OneLang.One.Ast.Expressions as exprs
import OneLang.One.Ast.Statements as stats
import OneLang.One.Ast.Types as types
import OneLang.One.Ast.References as refs
import OneLang.One.ErrorManager as errorMan
import OneLang.One.ITransformer as iTrans
import OneLang.One.Ast.Interfaces as ints

class AstTransformer:
    def __init__(self, name):
        self.error_man = errorMan.ErrorManager()
        self.current_file = None
        self.current_interface = None
        self.current_method = None
        self.current_closure = None
        self.current_statement = None
        self.name = name
    
    def visit_attributes_and_trivia(self, node):
        pass
    
    def visit_type(self, type):
        if isinstance(type, astTypes.ClassType) or isinstance(type, astTypes.InterfaceType) or isinstance(type, astTypes.UnresolvedType):
            type2 = type
            type2.type_arguments = list(map(lambda x: self.visit_type(x), type2.type_arguments))
        elif isinstance(type, astTypes.LambdaType):
            for mp in type.parameters:
                self.visit_method_parameter(mp)
            type.return_type = self.visit_type(type.return_type)
        return type
    
    def visit_identifier(self, id):
        return id
    
    def visit_variable(self, variable):
        if variable.type != None:
            variable.type = self.visit_type(variable.type)
        return variable
    
    def visit_variable_with_initializer(self, variable):
        self.visit_variable(variable)
        if variable.initializer != None:
            variable.initializer = self.visit_expression(variable.initializer)
        return variable
    
    def visit_variable_declaration(self, stmt):
        self.visit_variable_with_initializer(stmt)
        return stmt
    
    def visit_unknown_statement(self, stmt):
        self.error_man.throw(f'''Unknown statement type''')
        return stmt
    
    def visit_statement(self, stmt):
        self.current_statement = stmt
        self.visit_attributes_and_trivia(stmt)
        if isinstance(stmt, stats.ReturnStatement):
            if stmt.expression != None:
                stmt.expression = self.visit_expression(stmt.expression)
        elif isinstance(stmt, stats.ExpressionStatement):
            stmt.expression = self.visit_expression(stmt.expression)
        elif isinstance(stmt, stats.IfStatement):
            stmt.condition = self.visit_expression(stmt.condition)
            stmt.then = self.visit_block(stmt.then)
            if stmt.else_ != None:
                stmt.else_ = self.visit_block(stmt.else_)
        elif isinstance(stmt, stats.ThrowStatement):
            stmt.expression = self.visit_expression(stmt.expression)
        elif isinstance(stmt, stats.VariableDeclaration):
            return self.visit_variable_declaration(stmt)
        elif isinstance(stmt, stats.WhileStatement):
            stmt.condition = self.visit_expression(stmt.condition)
            stmt.body = self.visit_block(stmt.body)
        elif isinstance(stmt, stats.DoStatement):
            stmt.condition = self.visit_expression(stmt.condition)
            stmt.body = self.visit_block(stmt.body)
        elif isinstance(stmt, stats.ForStatement):
            if stmt.item_var != None:
                self.visit_variable_with_initializer(stmt.item_var)
            stmt.condition = self.visit_expression(stmt.condition)
            stmt.incrementor = self.visit_expression(stmt.incrementor)
            stmt.body = self.visit_block(stmt.body)
        elif isinstance(stmt, stats.ForeachStatement):
            self.visit_variable(stmt.item_var)
            stmt.items = self.visit_expression(stmt.items)
            stmt.body = self.visit_block(stmt.body)
        elif isinstance(stmt, stats.TryStatement):
            stmt.try_body = self.visit_block(stmt.try_body)
            if stmt.catch_body != None:
                self.visit_variable(stmt.catch_var)
                stmt.catch_body = self.visit_block(stmt.catch_body)
            if stmt.finally_body != None:
                stmt.finally_body = self.visit_block(stmt.finally_body)
        elif isinstance(stmt, stats.BreakStatement):
            pass
        elif isinstance(stmt, stats.UnsetStatement):
            stmt.expression = self.visit_expression(stmt.expression)
        elif isinstance(stmt, stats.ContinueStatement):
            pass
        else:
            return self.visit_unknown_statement(stmt)
        return stmt
    
    def visit_block(self, block):
        block.statements = list(map(lambda x: self.visit_statement(x), block.statements))
        return block
    
    def visit_template_string(self, expr):
        i = 0
        
        while i < len(expr.parts):
            part = expr.parts[i]
            if not part.is_literal:
                part.expression = self.visit_expression(part.expression)
            i = i + 1
        return expr
    
    def visit_unknown_expression(self, expr):
        self.error_man.throw(f'''Unknown expression type''')
        return expr
    
    def visit_lambda(self, lambda_):
        prev_closure = self.current_closure
        self.current_closure = lambda_
        self.visit_method_base(lambda_)
        self.current_closure = prev_closure
        return lambda_
    
    def visit_variable_reference(self, var_ref):
        return var_ref
    
    def visit_expression(self, expr):
        if isinstance(expr, exprs.BinaryExpression):
            expr.left = self.visit_expression(expr.left)
            expr.right = self.visit_expression(expr.right)
        elif isinstance(expr, exprs.NullCoalesceExpression):
            expr.default_expr = self.visit_expression(expr.default_expr)
            expr.expr_if_null = self.visit_expression(expr.expr_if_null)
        elif isinstance(expr, exprs.UnresolvedCallExpression):
            expr.func = self.visit_expression(expr.func)
            expr.type_args = list(map(lambda x: self.visit_type(x), expr.type_args))
            expr.args = list(map(lambda x: self.visit_expression(x), expr.args))
        elif isinstance(expr, exprs.UnresolvedMethodCallExpression):
            expr.object = self.visit_expression(expr.object)
            expr.type_args = list(map(lambda x: self.visit_type(x), expr.type_args))
            expr.args = list(map(lambda x: self.visit_expression(x), expr.args))
        elif isinstance(expr, exprs.ConditionalExpression):
            expr.condition = self.visit_expression(expr.condition)
            expr.when_true = self.visit_expression(expr.when_true)
            expr.when_false = self.visit_expression(expr.when_false)
        elif isinstance(expr, exprs.Identifier):
            return self.visit_identifier(expr)
        elif isinstance(expr, exprs.UnresolvedNewExpression):
            self.visit_type(expr.cls_)
            expr.args = list(map(lambda x: self.visit_expression(x), expr.args))
        elif isinstance(expr, exprs.NewExpression):
            self.visit_type(expr.cls_)
            expr.args = list(map(lambda x: self.visit_expression(x), expr.args))
        elif isinstance(expr, exprs.TemplateString):
            return self.visit_template_string(expr)
        elif isinstance(expr, exprs.ParenthesizedExpression):
            expr.expression = self.visit_expression(expr.expression)
        elif isinstance(expr, exprs.UnaryExpression):
            expr.operand = self.visit_expression(expr.operand)
        elif isinstance(expr, exprs.PropertyAccessExpression):
            expr.object = self.visit_expression(expr.object)
        elif isinstance(expr, exprs.ElementAccessExpression):
            expr.object = self.visit_expression(expr.object)
            expr.element_expr = self.visit_expression(expr.element_expr)
        elif isinstance(expr, exprs.ArrayLiteral):
            expr.items = list(map(lambda x: self.visit_expression(x), expr.items))
        elif isinstance(expr, exprs.MapLiteral):
            for item in expr.items:
                item.value = self.visit_expression(item.value)
        elif isinstance(expr, exprs.StringLiteral):
            pass
        elif isinstance(expr, exprs.BooleanLiteral):
            pass
        elif isinstance(expr, exprs.NumericLiteral):
            pass
        elif isinstance(expr, exprs.NullLiteral):
            pass
        elif isinstance(expr, exprs.RegexLiteral):
            pass
        elif isinstance(expr, exprs.CastExpression):
            expr.new_type = self.visit_type(expr.new_type)
            expr.expression = self.visit_expression(expr.expression)
        elif isinstance(expr, exprs.InstanceOfExpression):
            expr.expr = self.visit_expression(expr.expr)
            expr.check_type = self.visit_type(expr.check_type)
        elif isinstance(expr, exprs.AwaitExpression):
            expr.expr = self.visit_expression(expr.expr)
        elif isinstance(expr, types.Lambda):
            return self.visit_lambda(expr)
        elif isinstance(expr, refs.ClassReference):
            pass
        elif isinstance(expr, refs.EnumReference):
            pass
        elif isinstance(expr, refs.ThisReference):
            pass
        elif isinstance(expr, refs.StaticThisReference):
            pass
        elif isinstance(expr, refs.MethodParameterReference):
            return self.visit_variable_reference(expr)
        elif isinstance(expr, refs.VariableDeclarationReference):
            return self.visit_variable_reference(expr)
        elif isinstance(expr, refs.ForVariableReference):
            return self.visit_variable_reference(expr)
        elif isinstance(expr, refs.ForeachVariableReference):
            return self.visit_variable_reference(expr)
        elif isinstance(expr, refs.CatchVariableReference):
            return self.visit_variable_reference(expr)
        elif isinstance(expr, refs.GlobalFunctionReference):
            pass
        elif isinstance(expr, refs.SuperReference):
            pass
        elif isinstance(expr, refs.InstanceFieldReference):
            expr.object = self.visit_expression(expr.object)
            return self.visit_variable_reference(expr)
        elif isinstance(expr, refs.InstancePropertyReference):
            expr.object = self.visit_expression(expr.object)
            return self.visit_variable_reference(expr)
        elif isinstance(expr, refs.StaticFieldReference):
            return self.visit_variable_reference(expr)
        elif isinstance(expr, refs.StaticPropertyReference):
            return self.visit_variable_reference(expr)
        elif isinstance(expr, refs.EnumMemberReference):
            pass
        elif isinstance(expr, exprs.StaticMethodCallExpression):
            expr.type_args = list(map(lambda x: self.visit_type(x), expr.type_args))
            expr.args = list(map(lambda x: self.visit_expression(x), expr.args))
        elif isinstance(expr, exprs.GlobalFunctionCallExpression):
            expr.args = list(map(lambda x: self.visit_expression(x), expr.args))
        elif isinstance(expr, exprs.InstanceMethodCallExpression):
            expr.object = self.visit_expression(expr.object)
            expr.type_args = list(map(lambda x: self.visit_type(x), expr.type_args))
            expr.args = list(map(lambda x: self.visit_expression(x), expr.args))
        elif isinstance(expr, exprs.LambdaCallExpression):
            expr.args = list(map(lambda x: self.visit_expression(x), expr.args))
        else:
            return self.visit_unknown_expression(expr)
        return expr
    
    def visit_method_parameter(self, method_parameter):
        self.visit_attributes_and_trivia(method_parameter)
        self.visit_variable_with_initializer(method_parameter)
    
    def visit_method_base(self, method):
        for item in method.parameters:
            self.visit_method_parameter(item)
        
        if method.body != None:
            method.body = self.visit_block(method.body)
    
    def visit_method(self, method):
        self.current_method = method
        self.current_closure = method
        self.visit_attributes_and_trivia(method)
        self.visit_method_base(method)
        method.returns = self.visit_type(method.returns)
        self.current_closure = None
        self.current_method = None
    
    def visit_global_function(self, func):
        self.visit_method_base(func)
        func.returns = self.visit_type(func.returns)
    
    def visit_constructor(self, constructor):
        self.current_method = constructor
        self.current_closure = constructor
        self.visit_attributes_and_trivia(constructor)
        self.visit_method_base(constructor)
        self.current_closure = None
        self.current_method = None
    
    def visit_field(self, field):
        self.visit_attributes_and_trivia(field)
        self.visit_variable_with_initializer(field)
    
    def visit_property(self, prop):
        self.visit_attributes_and_trivia(prop)
        self.visit_variable(prop)
        if prop.getter != None:
            prop.getter = self.visit_block(prop.getter)
        if prop.setter != None:
            prop.setter = self.visit_block(prop.setter)
    
    def visit_interface(self, intf):
        self.current_interface = intf
        self.visit_attributes_and_trivia(intf)
        intf.base_interfaces = list(map(lambda x: self.visit_type(x), intf.base_interfaces))
        for field in intf.fields:
            self.visit_field(field)
        for method in intf.methods:
            self.visit_method(method)
        self.current_interface = None
    
    def visit_class(self, cls_):
        self.current_interface = cls_
        self.visit_attributes_and_trivia(cls_)
        if cls_.constructor_ != None:
            self.visit_constructor(cls_.constructor_)
        
        cls_.base_class = self.visit_type(cls_.base_class)
        cls_.base_interfaces = list(map(lambda x: self.visit_type(x), cls_.base_interfaces))
        for field in cls_.fields:
            self.visit_field(field)
        for prop in cls_.properties:
            self.visit_property(prop)
        for method in cls_.methods:
            self.visit_method(method)
        self.current_interface = None
    
    def visit_enum(self, enum_):
        self.visit_attributes_and_trivia(enum_)
        for value in enum_.values:
            self.visit_enum_member(value)
    
    def visit_enum_member(self, enum_member):
        pass
    
    def visit_import(self, imp):
        self.visit_attributes_and_trivia(imp)
    
    def visit_file(self, source_file):
        self.error_man.reset_context(self)
        self.current_file = source_file
        for imp in source_file.imports:
            self.visit_import(imp)
        for enum_ in source_file.enums:
            self.visit_enum(enum_)
        for intf in source_file.interfaces:
            self.visit_interface(intf)
        for cls_ in source_file.classes:
            self.visit_class(cls_)
        for func in source_file.funcs:
            self.visit_global_function(func)
        source_file.main_block = self.visit_block(source_file.main_block)
        self.current_file = None
    
    def visit_files(self, files):
        for file in files:
            self.visit_file(file)
    
    def visit_package(self, pkg):
        self.visit_files(pkg.files.values())